libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1650,3 +1671,216 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
+	m_returnInfo.reset();
+	successClause.block().accept(*this);
+	m_code << "}\n";
+
+	m_code << "default { // failure case\n";
+	handleCatch(_tryStatement.clauses());
+	m_code << "}\n";
+
+	return false;
+}
+
+string IRGeneratorForStatements::decodeReturnParameters()
+{
+	ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+
+	return Whiskers(R"(
+		<?dynamicReturnSize>
+			// copy dynamic return data out
+			returndatacopy(<pos>, 0, returndatasize())
+		</dynamicReturnSize>
+
+		// update freeMemoryPointer according to dynamic return size
+		mstore(<freeMemoryPointer>, add(<pos>, and(add(<returnSize>, 0x1f), not(0x1f))))
+
+		// decode return parameters from external try-call into <retVars>
+		let <retVars> := <abiDecode>(<pos>, add(<pos>, <returnSize>))
+	)")
+	("retVars", IRVariable(m_returnInfo->functionCall).commaSeparatedList())
+	("pos", m_returnInfo->returndataVariable)
+	("abiDecode", abi.tupleDecoder(m_returnInfo->returnTypes, true))
+	("freeMemoryPointer", to_string(CompilerUtils::freeMemoryPointer))
+	("dynamicReturnSize", m_returnInfo->dynamicReturnSize)
+	("returnSize",
+		m_context.evmVersion().supportsReturndata()
+			? "returndatasize()"s
+			: to_string(m_returnInfo->estimatedReturnSize)
+	).render();
+}
+
+void IRGeneratorForStatements::decodeTryCallReturnParameters(TryCatchClause const& _successClause, TypePointer _callReturnType)
+{
+	solAssert(m_returnInfo.has_value(), "");
+	verifyTrySuccessClause(_successClause, _callReturnType);
+	if (!_successClause.parameters())
+		return;
+
+	m_code << decodeReturnParameters();
+
+	size_t i = 0;
+	for (ASTPointer<VariableDeclaration> const& varDecl: _successClause.parameters()->parameters())
+	{
+		solAssert(varDecl, "");
+		define(
+			m_context.addLocalVariable(*varDecl),
+			IRVariable(m_returnInfo->functionCall).tupleComponent(i++)
+		);
+	}
+}
+
+void IRGeneratorForStatements::verifyTrySuccessClause(TryCatchClause const& _successClause, TypePointer _callReturnType)
+{
+	if (_successClause.parameters())
+	{
+		vector<TypePointer> exprTypes{_callReturnType};
+		if (auto tupleType = dynamic_cast<TupleType const*>(exprTypes.front()))
+			exprTypes = tupleType->components();
+		vector<ASTPointer<VariableDeclaration>> const& params = _successClause.parameters()->parameters();
+		solAssert(exprTypes.size() == params.size(), "");
+		for (size_t i = 0; i < exprTypes.size(); ++i)
+			solAssert(params[i] && exprTypes[i] && *params[i]->annotation().type == *exprTypes[i], "");
+	}
+}
+
+void IRGeneratorForStatements::handleCatch(vector<ASTPointer<TryCatchClause>> const& _catchClauses)
+{
+	auto const [structured, fallback] = extractCatchClauses(_catchClauses);
+
+	if (structured)
+		handleCatchStructured(*structured, fallback);
+	else if (fallback)
+		handleCatchFallback(*fallback);
+	else
+		m_code << rethrowCode();
+}
+
+tuple<TryCatchClause const*, TryCatchClause const*> IRGeneratorForStatements::extractCatchClauses(
```

Comment:
This is cluttering the code generation with code that's entirely unrelated to code generation.
I think there should rather be a member function in ``TryStatement`` that gives you these - and then you can directly use them to dispatch in place of ``handleCatch(_tryStatement.clauses());`` in ``IRGeneratorForStatements::visit(TryStatement const& _tryStatement)``.